stdlib/REPL/src/REPLCompletions.jl

```diff
@@ -1550,44 +1326,52 @@ function shell_completions(string, pos, hint::Bool=false)
-        was_expanded = expanded[3]
-        if was_expanded
-            path = (only(expanded[1])::PathCompletion).path
-            # If tab press, ispath and user expansion available, return it now
-            # otherwise see if we can complete the path further before returning with expanded ~
-            !hint && ispath(path) && return expanded::Completions
-        end
+        paths, success = complete_path_string(path, hint; use_envpath, shell_escape=true, cmd_escape)
+        return paths, r, success
+    end
+    return Completion[], 1:0, false
+end
 
-        paths, dir, success = complete_path(path, use_envpath=use_envpath, shell_escape=true, contract_user=was_expanded)
-
-        if success && !isempty(dir)
-            let dir = do_shell_escape(dir)
-                # if escaping of dir matches scs prefix, remove that from the completions
-                # otherwise make it the whole completion
-                partial = string[last_arg_start:pos]
-                if endswith(dir, "/") && startswith(partial, dir)
-                    r = (last_arg_start + sizeof(dir)):pos
-                elseif startswith(partial, dir * "/")
-                    r = nextind(string, last_arg_start + sizeof(dir)):pos
-                else
-                    map!(paths, paths) do c::PathCompletion
-                        return PathCompletion(dir * "/" * c.path)
-                    end
-                end
-            end
+function complete_path_string(path, hint::Bool=false;
+                              shell_escape::Bool=false,
+                              cmd_escape::Bool=false,
+                              string_escape::Bool=false,
+                              kws...)
+    # Expand "~" and remember if we expanded it.
+    local expanded
+    try
+        let p = expanduser(path)
+            expanded = path != p
+            path = p
         end
-        # if ~ was expanded earlier and the incomplete string isn't a path
-        # return the path with contracted user to match what the hint shows. Otherwise expand ~
-        # i.e. require two tab presses to expand user
-        if was_expanded && !ispath(path)
-            map!(paths, paths) do c::PathCompletion
-                PathCompletion(contractuser(c.path))
-            end
-        end
-        return paths, r, success
+    catch e
+        e isa ArgumentError || rethrow()
+        expanded = false
+    end
+
+    function escape(p)
+        shell_escape && (p = do_shell_escape(p))
+        string_escape && (p = do_string_escape(p))
+        cmd_escape && (p = do_cmd_escape(p))
+        p
+    end
+
+    paths, dir, success = complete_path(path; contract_user=expanded, shell_escape, cmd_escape, string_escape, kws...)
+    dir = dir == "" ? dir : escape(dir)
+
+    # Expand '~' if the user hits TAB after exhausting completions (either
+    # because we have found an existing file, or there is no such file).
+    full_path = ispath(path) || isempty(paths)
```

Comment:
I found two cases that can cause `ispath` to error, so I think we should make a `safe_ispath`

Something like this? Or just never throw and just return false?
```
function safe_ispath(path::AbstractString)
    try
        return ispath(path)
    catch err
        if err isa Base.IOError # i.e. name too long (ENAMETOOLONG)
            return false
        elseif err isa Base.ArgumentError && occursin("embedded NULs", err.msg)
            return false
        else
            rethrow()
        end
    end
end
```

Reply:
I'll avoid modifying Base for the time being because we already convert some `access(2)` return values to `false`.  It feels a little hacky to keep adding exceptions here. https://github.com/JuliaLang/julia/blob/bc98abcfbe392d2a0a09ed631b85f4675a34b380/base/stat.jl#L376-L379

Of the errors that could be returned for `F_OK` that we might want to handle, it looks like we're missing `EACCES`, `EIO`, `ELOOP`, and `ENAMETOOLONG`.